derive.h

```diff
@@ -0,0 +1,311 @@
+	}
+protected:
+	friend class BinaryExpression;
+	virtual std::optional<char> opt_var() const { return std::nullopt; }
+	Expression() = default;
+};
+
+ostream& operator<<(ostream& os, const Expression& ex) {
+	return ex.printTo(os);
+}
+
+class CoefExpression : public Expression {
+public:
+	CoefExpression(double coef): m_coef(coef) 
+	{ }
+	
+	bool is_zero() const override { return coef() == 0.; }
+	shared_ptr<Expression> derivative() const override {
+		return shared_ptr<CoefExpression>(new CoefExpression{0});
+	}
+	ostream& printTo(ostream& os) const override {
+		os << coef();
+		return os;
+	}
+	double coef() const { return m_coef;}
+	friend class BinaryExpression;
+protected:
+	optional<char> opt_var() const override { return std::nullopt; }
+	double m_coef;
+};
+
+
+class XExpression : public Expression {
+public:
+	ostream& printTo(ostream& os) const override {
+		os << x();
+		return os;
+	}
+
+	bool is_zero() const override { return false; }
+	shared_ptr<Expression> derivative() const override {
+		return	shared_ptr<CoefExpression>(new CoefExpression{1});
+	}
+	char x() const { return m_x; }
+	XExpression(char x) { m_x = x; }
+	friend class BinaryExpression;
+protected:
+	optional<char> opt_var() const override { return optional<char>{m_x}; }
+	char m_x;
+};
+
+class BinaryExpression;
+shared_ptr<Expression> make_shared_expression(BinaryExpression* ptr);
+shared_ptr<Expression> make_shared_expression(CoefExpression* ptr);
+shared_ptr<Expression> multiply(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs);
+shared_ptr<Expression> add(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs);
+shared_ptr<Expression> pow(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs);
+shared_ptr<Expression> subtract(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs);
+
+class BinaryExpression : public Expression {
+public:
+	ostream& printTo(ostream& os) const override {
+		os << "(";
+		lhs->printTo(os);
+	 	os << operation;
+		rhs->printTo(os);
+		os << ")";
+		return os;
+	}
+
+	bool is_zero() const override {
+		switch (operation) {
+			case plus: 
+				return lhs->is_zero() && rhs->is_zero();
+			case minus: 
+				return lhs->is_zero() && rhs->is_zero();
+			case mult: 
+				return lhs->is_zero() || rhs->is_zero();
+			case power:
+				return lhs->is_zero();
+		}
+		assert(0);
+		return false;
+	}
+
+	
+
+	BinaryExpression(shared_ptr<Expression> a_lhs, shared_ptr<Expression> a_rhs, Operation a_operation):
+		lhs(a_lhs),
+		rhs(a_rhs),
+		operation(a_operation)
+	{ 
+		opt_var();
+	}
+	
+	shared_ptr<Expression> derivative() const override {
+		switch (operation) {
+			case plus: 
+				return add(lhs->derivative(), rhs->derivative());
+			case minus: 
+				return subtract(lhs->derivative(), rhs->derivative());
+			case mult: 
+				return add(
+					multiply(lhs->derivative(), rhs),
+					multiply(rhs->derivative(), lhs)
+				);
+			case power: 
+				{
```

Comment:
Why is this new block/

Reply:
Because I am using a variable inside and the compiler does not agree to it if it is not in a separate block.